### ISSUE TITLE
Need AsyncAbstractBufferedFile or guidence

### ISSUE THREAD

**Issue by awm33** [intake/filesystem_spec#714]
Context - I'm looking to use fsspec in async backend service, so I'm interested in implementing async versions (or replacements) of SFTP, FTP, and potentially others. However, for it to work well in this type of environment without blocking, it would need to support async in depth.

It looks like AsyncFileSystem would provide a standard interface for working with async versions of the functions (using an underscored version of the function ex `_ls()`), but when working with files, there doesn't seem to be a standard. The "http" protocol implementation includes `async_fetch_all` and `async_fetch_range ` functions but S3FS does not, so it appears non-standard.

How do we see making file level functions like `read` work without blocking? Do we need a file level class similar to `AsyncFileSystem`?

**Comment by martindurant** [comment]
You are correct that fsspec does not support async files at all. One might argue that the random-access nature of files is contrary to any async usage pattern (you envision one coroutine accessing one file, but would not necessarily be the case).

You are welcome to try to implement an async file interface, but I suspect it would be difficult, if we were to support the synchronous API too - which is a requirement. Perhaps when `_open` happens for a filesystem where `asynchronous=True`, you get a different class instantiated...
